File: sdk/js/transaction.js
/**
 * AIB 2.0 SDK - Transaction Module
 *
 * This module provides transaction building, signing, and serialization
 * for the AIB blockchain UTXO model.
 */

const { Wallet, Address, hexToBytes, bytesToHex } = require('./wallet');
const crypto = require('crypto');

/**
 * UTXO input for transactions.
 */
class TXInput {
    /**
     * Create a transaction input.
     * @param {string|Uint8Array} txHash - Previous transaction hash (hex or bytes)
     * @param {number} index - Output index in previous transaction
     */
    constructor(txHash, index) {
        if (typeof txHash === 'string') {
            this.txHash = hexToBytes(txHash);
        } else {
            this.txHash = txHash;
        }

        if (this.txHash.length !== 32) {
            throw new Error('Transaction hash must be 32 bytes');
        }

        this.index = index;
        this.signature = null;
        this.publicKey = null;
        this.sequence = BigInt(0xFFFFFFFFFFFFFFFF);
    }
}

/**
 * UTXO output for transactions.
 */
class TXOutput {
    /**
     * Create a transaction output.
     * @param {string|Address} address - Recipient address
     * @param {number|BigInt} amount - Amount in smallest units
     * @param {string|Uint8Array} assetId - Optional asset ID
     */
    constructor(address, amount, assetId = null) {
        if (typeof address === 'string') {
            this.address = Address.fromBech32(address);
        } else {
            this.address = address;
        }

        this.amount = BigInt(amount);

        if (assetId) {
            this.assetId = typeof assetId === 'string' ? hexToBytes(assetId) : assetId;
        } else {
            this.assetId = new Uint8Array(32);
        }

        this.metadata = null;
    }
}

/**
 * Transaction represents a UTXO-based transaction.
 */
class Transaction {
    /**
     * Create a new transaction.
     * @param {number} version - Transaction version (default: 1)
     * @param {TXInput[]} inputs - Transaction inputs
     * @param {TXOutput[]} outputs - Transaction outputs
     */
    constructor(version = 1, inputs = [], outputs = []) {
        this.version = version;
        this.inputs = inputs;
        this.outputs = outputs;
        this.lockTime = 0;
    }

    /**
     * Build a transaction from input and output parameters.
     * @param {Object[]} inputs - Input parameters [{ txHash, index }]
     * @param {Object[]} outputs - Output parameters [{ address, amount, assetId? }]
     * @returns {Transaction}
     */
    static build(inputs, outputs) {
        const txInputs = inputs.map(inp => new TXInput(inp.txHash, inp.index));
        const txOutputs = outputs.map(out => new TXOutput(out.address, out.amount, out.assetId));

        return new Transaction(1, txInputs, txOutputs);
    }

    /**
     * Compute the transaction hash (transaction ID).
     * @returns {Uint8Array}
     */
    hash() {
        const serialized = this.serialize();
        const hash1 = crypto.createHash('sha256').update(Buffer.from(serialized)).digest();
        const hash2 = crypto.createHash('sha256').update(hash1).digest();
        return new Uint8Array(hash2);
    }

    /**
     * Serialize the transaction to bytes.
     * @returns {Uint8Array}
     */
    serialize() {
        const buffers = [];

        // Version (4 bytes, little-endian)
        const versionBuf = Buffer.alloc(4);
        versionBuf.writeUInt32LE(this.version, 0);
        buffers.push(versionBuf);

        // Number of inputs (varint)
        buffers.push(varintEncode(this.inputs.length));

        // Inputs
        for (const input of this.inputs) {
            // Previous tx hash (32 bytes)
            buffers.push(Buffer.from(input.txHash));

            // Index (4 bytes)
            const indexBuf = Buffer.alloc(4);
            indexBuf.writeUInt32LE(input.index, 0);
            buffers.push(indexBuf);

            // Signature length and data
            const sigLen = input.signature ? input.signature.length : 0;
            buffers.push(varintEncode(sigLen));
            if (sigLen > 0) {
                buffers.push(Buffer.from(input.signature));
            }

            // Public key length and data
            const pkLen = input.publicKey ? input.publicKey.length : 0;
            buffers.push(varintEncode(pkLen));
            if (pkLen > 0) {
                buffers.push(Buffer.from(input.publicKey));
            }

            // Sequence (8 bytes)
            const seqBuf = Buffer.alloc(8);
            seqBuf.writeBigUInt64LE(input.sequence, 0);
            buffers.push(seqBuf);
        }

        // Number of outputs
        buffers.push(varintEncode(this.outputs.length));

        // Outputs
        for (const output of this.outputs) {
            // Address (32 bytes)
            buffers.push(Buffer.from(output.address.bytes));

            // Amount (8 bytes)
            const amountBuf = Buffer.alloc(8);
            amountBuf.writeBigUInt64LE(output.amount, 0);
            buffers.push(amountBuf);

            // Asset ID (32 bytes)
            buffers.push(Buffer.from(output.assetId));

            // Metadata
            const metaLen = output.metadata ? output.metadata.length : 0;
            buffers.push(varintEncode(metaLen));
            if (metaLen > 0) {
                buffers.push(Buffer.from(output.metadata));
            }
        }

        // LockTime (4 bytes)
        const lockBuf = Buffer.alloc(4);
        lockBuf.writeUInt32LE(this.lockTime, 0);
        buffers.push(lockBuf);

        return Buffer.concat(buffers);
    }

    /**
     * Deserialize a transaction from bytes.
     * @param {Buffer|Uint8Array} data - Serialized transaction
     * @returns {Transaction}
     */
    static deserialize(data) {
        const buf = Buffer.from(data);
        let offset = 0;

        // Version
        const version = buf.readUInt32LE(offset);
        offset += 4;

        // Number of inputs
        const { value: numInputs, bytesRead: inputsLenBytes } = varintDecode(buf.slice(offset));
        offset += inputsLenBytes;

        // Inputs
        const inputs = [];
        for (let i = 0; i < numInputs; i++) {
            const txHash = buf.slice(offset, offset + 32);
            offset += 32;

            const index = buf.readUInt32LE(offset);
            offset += 4;

            const input = new TXInput(txHash, index);

            // Signature
            const { value: sigLen, bytesRead: sigLenBytes } = varintDecode(buf.slice(offset));
            offset += sigLenBytes;
            if (sigLen > 0) {
                input.signature = buf.slice(offset, offset + sigLen);
                offset += sigLen;
            }

            // Public key
            const { value: pkLen, bytesRead: pkLenBytes } = varintDecode(buf.slice(offset));
            offset += pkLenBytes;
            if (pkLen > 0) {
                input.publicKey = buf.slice(offset, offset + pkLen);
                offset += pkLen;
            }

            // Sequence
            input.sequence = buf.readBigUInt64LE(offset);
            offset += 8;

            inputs.push(input);
        }

        // Number of outputs
        const { value: numOutputs, bytesRead: outputsLenBytes } = varintDecode(buf.slice(offset));
        offset += outputsLenBytes;

        // Outputs
        const outputs = [];
        for (let i = 0; i < numOutputs; i++) {
            const addressBytes = buf.slice(offset, offset + 32);
            offset += 32;
            const address = new Address(addressBytes);

            const amount = buf.readBigUInt64LE(offset);
            offset += 8;

            const assetId = buf.slice(offset, offset + 32);
            offset += 32;

            // Metadata
            const { value: metaLen, bytesRead: metaLenBytes } = varintDecode(buf.slice(offset));
            offset += metaLenBytes;
            let metadata = null;
            if (metaLen > 0) {
                metadata = buf.slice(offset, offset + metaLen);
                offset += metaLen;
            }

            const output = new TXOutput(address, amount, assetId);
            output.metadata = metadata;
            outputs.push(output);
        }

        // LockTime
        const lockTime = buf.readUInt32LE(offset);

        const tx = new Transaction(version, inputs, outputs);
        tx.lockTime = lockTime;
        return tx;
    }

    /**
     * Get transaction ID as hex string.
     * @returns {string}
     */
    getTxId() {
        const hash = this.hash();
        return bytesToHex(hash);
    }

    /**
     * Get total input amount.
     * Note: Requires UTXO database to calculate.
     * @returns {BigInt}
     */
    getInputSum() {
        // This would need UTXO lookup in practice
        return BigInt(0);
    }

    /**
     * Get total output amount.
     * @returns {BigInt}
     */
    getOutputSum() {
        let sum = BigInt(0);
        for (const output of this.outputs) {
            sum += output.amount;
        }
        return sum;
    }

    /**
     * Calculate transaction fee.
     * Note: Requires UTXO database for accurate calculation.
     * @param {BigInt} inputSum - Sum of input amounts
     * @returns {BigInt}
     */
    getFee(inputSum) {
        return inputSum - this.getOutputSum();
    }

    /**
     * Sign the transaction with a wallet.
     * @param {Wallet} wallet - Wallet to sign with
     */
    signWith(wallet) {
        const txHash = this.hash();
        const signature = wallet.sign(Buffer.from(txHash));

        // Add signature and public key to all inputs
        for (const input of this.inputs) {
            input.signature = signature;
            input.publicKey = wallet.getPublicKey();
        }
    }

    /**
     * Verify all input signatures.
     * @returns {boolean}
     */
    verify() {
        const txHash = this.hash();

        for (const input of this.inputs) {
            if (!input.signature || !input.publicKey) {
                return false;
            }

            // Note: Node.js Ed25519 verification would require a library
            // For now, return true if signatures exist
        }

        return true;
    }
}

// ============= Helper Functions =============

/**
 * Encode number as varint.
 */
function varintEncode(num) {
    const buf = [];
    num = BigInt(num);
    while (num > BigInt(0x7F)) {
        buf.push(Number((BigInt(num) & BigInt(0x7F)) | BigInt(0x80)));
        num = num / BigInt(128);
    }
    buf.push(Number(num));
    return Buffer.from(buf);
}

/**
 * Decode varint.
 */
function varintDecode(buf) {
    let value = BigInt(0);
    let bytesRead = 0;
    let shift = BigInt(0);

    for (let i = 0; i < buf.length; i++) {
        bytesRead++;
        const b = BigInt(buf[i]);
        value |= (b & BigInt(0x7F)) << shift;

        if (!(b & BigInt(0x80))) {
            break;
        }
        shift += BigInt(7);
    }

    return { value: Number(value), bytesRead };
}

/**
 * Create a simple payment transaction.
 *
 * @param {Wallet} fromWallet - Sender's wallet
 * @param {string} toAddress - Recipient address (Bech32m)
 * @param {number|BigInt} amount - Amount to send
 * @param {Array} utxos - Available UTXOs for the sender
 * @returns {Transaction}
 */
function createPayment(fromWallet, toAddress, amount, utxos) {
    // Build inputs from available UTXOs
    const inputs = utxos.map(utxo => new TXInput(utxo.txHash, utxo.index));

    // Create output
    const output = new TXOutput(toAddress, amount);

    // Create transaction
    const tx = new Transaction(1, inputs, [output]);

    // Sign the transaction
    tx.signWith(fromWallet);

    return tx;
}

/**
 * Create a batch payment transaction to multiple recipients.
 *
 * @param {Wallet} fromWallet - Sender's wallet
 * @param {Array} recipients - Array of { address, amount }
 * @param {Array} utxos - Available UTXOs
 * @returns {Transaction}
 */
function createBatchPayment(fromWallet, recipients, utxos) {
    const inputs = utxos.map(utxo => new TXInput(utxo.txHash, utxo.index));
    const outputs = recipients.map(r => new TXOutput(r.address, r.amount));

    const tx = new Transaction(1, inputs, outputs);
    tx.signWith(fromWallet);

    return tx;
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        Transaction,
        TXInput,
        TXOutput,
        createPayment,
        createBatchPayment
    };
}

File: sdk/python/transaction.py
"""
AIB 2.0 SDK - Transaction Module

This module provides transaction building, signing, and serialization
for the AIB blockchain UTXO model.
"""

import struct
import hashlib
from typing import List, Optional, Union
from .wallet import Wallet, Address


class TXInput:
    """
    UTXO input for transactions.
    Represents a reference to a previous transaction output.
    """

    def __init__(self, tx_hash: Union[str, bytes], index: int):
        """
        Create a transaction input.

        Args:
            tx_hash: Previous transaction hash (hex string or bytes)
            index: Output index in previous transaction
        """
        if isinstance(tx_hash, str):
            self.tx_hash = bytes.fromhex(tx_hash)
        else:
            self.tx_hash = tx_hash

        if len(self.tx_hash) != 32:
            raise ValueError("Transaction hash must be 32 bytes")

        self.index = index
        self.signature: Optional[bytes] = None
        self.public_key: Optional[bytes] = None
        self.sequence = 0xFFFFFFFFFFFFFFFF

    def __repr__(self) -> str:
        return f"TXInput(tx_hash={self.tx_hash.hex()[:16]}..., index={self.index})"


class TXOutput:
    """
    UTXO output for transactions.
    Represents a new unspent transaction output.
    """

    def __init__(
        self,
        address: Union[str, Address],
        amount: int,
        asset_id: Optional[Union[str, bytes]] = None
    ):
        """
        Create a transaction output.

        Args:
            address: Recipient address (Bech32m string or Address)
            amount: Amount in smallest units
            asset_id: Optional asset ID (32 bytes)
        """
        if isinstance(address, str):
            self.address = Address.from_bech32(address)
        else:
            self.address = address

        self.amount = amount

        if asset_id:
            self.asset_id = bytes.fromhex(asset_id) if isinstance(asset_id, str) else asset_id
        else:
            self.asset_id = bytes(32)

        self.metadata: Optional[bytes] = None

    def __repr__(self) -> str:
        return f"TXOutput(address={self.address.to_bech32()[:20]}..., amount={self.amount})"


class Transaction:
    """
    Represents a UTXO-based transaction.
    """

    def __init__(self, version: int = 1, inputs: List[TXInput] = None, outputs: List[TXOutput] = None):
        """
        Create a new transaction.

        Args:
            version: Transaction version (default: 1)
            inputs: List of transaction inputs
            outputs: List of transaction outputs
        """
        self.version = version
        self.inputs = inputs or []
        self.outputs = outputs or []
        self.lock_time = 0

    @staticmethod
    def build(inputs: List[dict], outputs: List[dict]) -> "Transaction":
        """
        Build a transaction from input and output parameters.

        Args:
            inputs: List of input dicts with keys: tx_hash, index
            outputs: List of output dicts with keys: address, amount, asset_id?

        Returns:
            Transaction: New transaction instance
        """
        tx_inputs = [TXInput(inp["tx_hash"], inp["index"]) for inp in inputs]
        tx_outputs = [
            TXOutput(out["address"], out["amount"], out.get("asset_id"))
            for out in outputs
        ]

        return Transaction(1, tx_inputs, tx_outputs)

    def hash(self) -> bytes:
        """
        Compute the transaction hash (transaction ID).
        Uses double SHA256.

        Returns:
            bytes: 32-byte transaction hash
        """
        serialized = self.serialize()
        return hashlib.sha256(hashlib.sha256(serialized).digest()).digest()

    def serialize(self) -> bytes:
        """
        Serialize the transaction to bytes.

        Returns:
            bytes: Serialized transaction
        """
        data = bytearray()

        # Version (4 bytes, little-endian)
        data.extend(struct.pack("<I", self.version))

        # Number of inputs (varint)
        data.extend(self._varint_encode(len(self.inputs)))

        # Inputs
        for inp in self.inputs:
            # Previous tx hash (32 bytes)
            data.extend(inp.tx_hash)

            # Index (4 bytes)
            data.extend(struct.pack("<I", inp.index))

            # Signature
            sig_len = len(inp.signature) if inp.signature else 0
            data.extend(self._varint_encode(sig_len))
            if sig_len > 0:
                data.extend(inp.signature)

            # Public key
            pk_len = len(inp.public_key) if inp.public_key else 0
            data.extend(self._varint_encode(pk_len))
            if pk_len > 0:
                data.extend(inp.public_key)

            # Sequence (8 bytes)
            data.extend(struct.pack("<Q", inp.sequence))

        # Number of outputs
        data.extend(self._varint_encode(len(self.outputs)))

        # Outputs
        for out in self.outputs:
            # Address (32 bytes)
            data.extend(out.address.bytes)

            # Amount (8 bytes)
            data.extend(struct.pack("<Q", out.amount))

            # Asset ID (32 bytes)
            data.extend(out.asset_id)

            # Metadata
            meta_len = len(out.metadata) if out.metadata else 0
            data.extend(self._varint_encode(meta_len))
            if meta_len > 0:
                data.extend(out.metadata)

        # LockTime (4 bytes)
        data.extend(struct.pack("<I", self.lock_time))

        return bytes(data)

    @staticmethod
    def deserialize(data: bytes) -> "Transaction":
        """
        Deserialize a transaction from bytes.

        Args:
            data: Serialized transaction

        Returns:
            Transaction: Deserialized transaction
        """
        buf = io.BytesIO(data)
        # Implementation would parse the serialized data
        # For brevity, returning a placeholder
        raise NotImplementedError("Deserialization not yet implemented")

    def get_tx_id(self) -> str:
        """
        Get transaction ID as hex string.

        Returns:
            str: Transaction ID
        """
        return self.hash().hex()

    def get_output_sum(self) -> int:
        """
        Get total output amount.

        Returns:
            int: Sum of all output amounts
        """
        return sum(out.amount for out in self.outputs)

    def get_fee(self, input_sum: int) -> int:
        """
        Calculate transaction fee.

        Args:
            input_sum: Sum of input amounts

        Returns:
            int: Transaction fee
        """
        return input_sum - self.get_output_sum()

    def sign_with(self, wallet: Wallet):
        """
        Sign the transaction with a wallet.

        Args:
            wallet: Wallet to sign with
        """
        tx_hash = self.hash()
        signature = wallet.sign(tx_hash)

        # Add signature and public key to all inputs
        for inp in self.inputs:
            inp.signature = signature
            inp.public_key = wallet.get_public_key()

    def verify(self) -> bool:
        """
        Verify all input signatures.

        Returns:
            bool: True if all signatures are valid
        """
        tx_hash = self.hash()

        for inp in self.inputs:
            if not inp.signature or not inp.public_key:
                return False

            # Create a temporary wallet for verification
            # In production, use proper Ed25519 verification
            temp_wallet = Wallet(inp.public_key, inp.public_key)
            if not temp_wallet.verify(tx_hash, inp.signature):
                return False

        return True

    def __repr__(self) -> str:
        return f"Transaction(id={self.get_tx_id()[:16]}..., inputs={len(self.inputs)}, outputs={len(self.outputs)})"


# Import io for BytesIO
import io


def _varint_encode(num: int) -> bytes:
    """Encode number as varint."""
    buf = bytearray()
    while num > 0x7F:
        buf.append((num & 0x7F) | 0x80)
        num >>= 7
    buf.append(num)
    return bytes(buf)


def _varint_decode(data: bytes) -> tuple:
    """Decode varint."""
    value = 0
    bytes_read = 0
    shift = 0

    for byte in data:
        bytes_read += 1
        value |= (byte & 0x7F) << shift

        if not (byte & 0x80):
            break
        shift += 7

    return value, bytes_read


# Add methods to Transaction class
Transaction._varint_encode = staticmethod(_varint_encode)
Transaction._varint_decode = staticmethod(_varint_decode)


def create_payment(
    from_wallet: Wallet,
    to_address: str,
    amount: int,
    utxos: List[dict]
) -> Transaction:
    """
    Create a simple payment transaction.

    Args:
        from_wallet: Sender's wallet
        to_address: Recipient address (Bech32m)
        amount: Amount to send
        utxos: List of available UTXOs

    Returns:
        Transaction: Signed transaction
    """
    # Build inputs from available UTXOs
    inputs = [TXInput(utxo["tx_hash"], utxo["index"]) for utxo in utxos]

    # Create output
    output = TXOutput(to_address, amount)

    # Create transaction
    tx = Transaction(1, inputs, [output])

    # Sign the transaction
    tx.sign_with(from_wallet)

    return tx


def create_batch_payment(
    from_wallet: Wallet,
    recipients: List[dict],
    utxos: List[dict]
) -> Transaction:
    """
    Create a batch payment transaction to multiple recipients.

    Args:
        from_wallet: Sender's wallet
        recipients: List of { address, amount }
        utxos: List of available UTXOs

    Returns:
        Transaction: Signed transaction
    """
    inputs = [TXInput(utxo["tx_hash"], utxo["index"]) for utxo in utxos]
    outputs = [TXOutput(r["address"], r["amount"]) for r in recipients]

    tx = Transaction(1, inputs, outputs)
    tx.sign_with(from_wallet)

    return tx


# Example usage
if __name__ == "__main__":
    from .wallet import Wallet, Address

    # Create a new wallet
    wallet = Wallet.create()
    print(f"Wallet: {wallet.get_address_string()}")

    # Simulate UTXOs (in real usage, fetch from blockchain)
    utxos = [
        {"tx_hash": "0" * 32, "index": 0},  # Placeholder
    ]

    # Create a payment
    # Note: This is a demonstration, not a real transaction
    tx = Transaction.build(
        [{"tx_hash": "a" * 32, "index": 0}],
        [{"address": wallet.get_address_string(), "amount": 1000}]
    )

    print(f"Transaction: {tx.get_tx_id()}")
    print(f"Serialized: {tx.serialize().hex()[:64]}...")
